src/perl/lib/wtsi_clarity/mq/consumer.pm

```diff
@@ -0,0 +1,278 @@
+package wtsi_clarity::mq::consumer;
+
+use Moose;
+use Carp;
+use Try::Tiny;
+use WTSI::DNAP::RabbitMQ::Client;
+use POSIX qw(strftime);
+use English qw(-no_match_vars);
+use AnyEvent;
+use Daemon::Control;
+
+with 'wtsi_clarity::util::configurable';
+
+our $VERSION = '0.0';
+
+my @REQUIRED_ATTRS  = qw/name queue pid_file stderr_file stdout_file/;
+
+my %REQUIRED_RO_STR = ( is => 'ro', isa => 'Str', required => 1, );
+
+my %CALLBACK_ATTRS  = (
+  is       => 'rw',
+  isa      => 'CodeRef',
+  required => 1,
+  lazy     => 1,
+);
+
+has [@REQUIRED_ATTRS] => ( %REQUIRED_RO_STR );
+
+has 'channel_name' => ( %REQUIRED_RO_STR, default => 'channel' . $PID, );
+
+has 'dlx'          => ( %REQUIRED_RO_STR, lazy => 1, builder => '_build_dlx' );
+
+has 'credentials' => (
+  is       => 'ro',
+  isa      => 'ArrayRef',
+  required => 1,
+  lazy     => 1,
+  builder  => '_build_credentials',
+);
+
+has 'daemon' => (
+  is      => 'ro',
+  isa     => 'Daemon::Control',
+  lazy    => 1,
+  builder => '_build_daemon',
+);
+
+has 'client' => (
+  is      => 'ro',
+  isa     => 'Object',
+  lazy    => 1,
+  builder => '_build_client',
+);
+
+has 'message_handler' => (
+  is       => 'ro',
+  isa      => 'WtsiClarityMessageHandler',
+  required => 1,
+);
+
+has 'on_startup'       => ( %CALLBACK_ATTRS, builder => '_build_on_startup', );
+has 'on_consume'       => ( %CALLBACK_ATTRS, builder => '_build_on_consume', );
+has 'on_consume_error' => ( %CALLBACK_ATTRS, builder => '_build_on_consume_error', );
+has 'on_consume_end'   => ( %CALLBACK_ATTRS, builder => '_build_default_callback', );
+has 'on_client_error'  => ( %CALLBACK_ATTRS, builder => '_build_on_client_error', );
+
+sub run {
+  my $self = shift;
+  $self->daemon->program_args(\@ARGV);
+  exit $self->daemon->run;
+}
+
+sub log_message {
+  my $self    = shift;
+  my $message = shift // q{};
+  print {*STDOUT} join q{ }, _now(), qq{$message\n} or croak 'Can not write to STDOUT';
+  return;
+}
+
+sub log_error_message {
+  my $self    = shift;
+  my $message = shift // q{};
+  print {*STDERR} join q{ }, _now(), qq{$message\n} or croak 'Can not write to STDERR';
+  return;
+}
+
+sub program {
+  my ($self, @args) = @_;
+
+  $self->on_startup->(@args);
+
+  $self->client->connect( @{$self->credentials} )
+               ->open_channel( name => $self->channel_name )
+               ->consume( channel => $self->channel_name, queue => $self->queue );
+
+  return AnyEvent->condvar->recv;
+}
+
+sub send_to_dlx {
```

Comment:
I think I may leave this one as public since, if a consumer decides to pass in their own `on_consume_error` callback, they may want to call this method inside that callback.